test/typing/with-types-infer.ts

```diff
@@ -1,22 +1,125 @@
-import { withTypes } from '../../riot'
+import { WithTypes, withTypes } from '../../riot'
 
-export const Component = withTypes({
-  onClick() {
-    this.update({ clicked: true })
-  },
+/**
+ * test: component can assign state to object
+ */
+export const Component1 = withTypes({
   onMounted() {
     this.state = { clicked: false }
-    this.onClick()
   },
 })
 
+/**
+ * test: component can't assign state to non Record object
+ */
 export const Component2 = withTypes({
   onClick() {
     this.update({ clicked: true })
   },
   onMounted() {
     //@ts-expect-error
     this.state = 2
+  },
+})
+
+/**
+ * test: component does infer this and its methods and properties
+ */
+export const Component3 = withTypes({
```

Comment:
Thank you for the following examples. I will try to fix these issues starting from those